src/hipoc/hipoc_program.cpp

```diff
@@ -58,31 +59,46 @@ namespace miopen {
 #if !MIOPEN_USE_COMGR
 namespace {
 
-inline bool ProduceCoV3()
+inline std::string GetCodeObjectVersionOptionImpl()
 {
-    // If env.var is set, then let's simply follow it.
-    if(IsEnabled(MIOPEN_DEBUG_OPENCL_ENFORCE_COV3{}))
-        return true;
-    if(IsDisabled(MIOPEN_DEBUG_OPENCL_ENFORCE_COV3{}))
-        return false;
-    // Otherwise, let's assume that OpenCL kernels shall be compiled to
-    // CO v3 format by default since ROCm 3.0. The simplest way to find out
-    // this right now is checking the HIP compiler version string.
-    return (HipCompilerVersion() >= external_tool_version_t{3, 0, -1});
+    auto code_object_version = miopen::Value(MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_VERSION{});
+    // Very basic syntax check:
+    if(code_object_version == 1 || code_object_version > 4)
+    {
+        MIOPEN_LOG_E("Bad MIOPEN_DEBUG_OPENCL_ENFORCE_CODE_OBJECT_VERSION, using default");
+        code_object_version = 0;
+    }
+
+    if(HipCompilerVersion() >= external_tool_version_t{4, 0, -1})
+    {
+        if(code_object_version == 0) // Env.var is unset.
+            return {};               // Rely on compiler's default.
+        else
+            return std::string("-mcode-object-version=") + std::to_string(code_object_version);
```

Comment:
@kzhuravl @t-tye Does this option really functional beginning from 4.0.0? CQE is having problems with 4.0.20425.

Reply:
Yes, this option is part of TargetID changes. Is CQE using the compiler that is built from the TargetID sources, or are they using some older BKC version of the compiler?

Note that internal builds may have started reporting they are 4.0 before the TargetID changes were committed. But internal testing should be using the latest mainline sources.